src/duplicates.rs

```diff
@@ -0,0 +1,204 @@
+						have_before = true;
+					}
+
+					if param.starts_with("after=") {
+						have_after = true;
+					}
+
+					if have_before && have_after {
+						// Successfully found both "before=" and "after=" in
+						// the GET. No need to investigate any of the remaining
+						// params.
+						break;
+					}
+				}
+
+				if have_after {
+					before = "t3_".to_owned();
+					before.push_str(&duplicates[0].id);
+				}
+
+				// Address potentially missing "after". If "before=" is in the
+				// GET, then "after" will be null in the JSON (see FIXME
+				// above).
+				if have_before {
+					// The next batch will need to start from one after the
+					// last post in the current batch.
+					after = "t3_".to_owned();
+					after.push_str(&duplicates[l - 1].id);
+
+					// Here is where things get terrible. Notice that we
+					// haven't set `before`. In order to do so, we will
+					// need to know if there is a batch that exists before
+					// this one, and doing so requires actually fetching the
+					// previous batch. In other words, we have to do yet one
+					// more GET to Reddit. There is no other way to determine
+					// whether or not to define `before`.
+					//
+					// We'll mitigate that by requesting at most one duplicate.
+					let new_path: String = format!("{}.json?before=t3_{}&limit=1&raw_json=1", req.uri().path(), &duplicates[0].id);
+					match json(new_path, true).await {
+						Ok(response) => {
+							if (&response[1]["data"]["children"]).as_array().unwrap_or(&Vec::new()).len() > 0 {
+								before = "t3_".to_owned();
+								before.push_str(&duplicates[0].id);
+							}
+						}
+						Err(msg) => {
+							// Abort entirely if we couldn't get the previous
+							// batch.
+							return error(req, msg).await;
+						}
+					}
+				} else {
+					after = response[1]["data"]["after"].as_str().unwrap_or_default().to_string();
+				}
+			}
+			let url = req.uri().to_string();
+
+			template(DuplicatesTemplate {
+				params: DuplicatesParams { before, after },
+				post,
+				duplicates,
+				prefs: Preferences::new(req),
+				url,
+				num_posts_filtered,
+				all_posts_filtered,
+			})
+		}
+
+		// Process error.
+		Err(msg) => {
+			if msg == "quarantined" {
+				let sub = req.param("sub").unwrap_or_default();
+				quarantine(req, sub)
+			} else {
+				error(req, msg).await
+			}
+		}
+	}
+}
+
+// DUPLICATES
+async fn parse_duplicates(json: &serde_json::Value, filters: &HashSet<String>) -> (Vec<Post>, u64, bool) {
+	let post_duplicates: &Vec<Value> = &(json["data"]["children"]).as_array().map_or(Vec::new(), ToOwned::to_owned);
+	let mut duplicates: Vec<Post> = Vec::new();
+
+	// Process each post and place them in the Vec<Post>.
+	for val in post_duplicates.iter() {
+		let post: Post = parse_post(&val).await;
+		duplicates.push(post);
+	}
+
+	let (num_posts_filtered, all_posts_filtered) = filter_posts(&mut duplicates, &filters);
+	return (duplicates, num_posts_filtered, all_posts_filtered);
```

Comment:
Return statement not necessary, you can leave `(duplicates, num_posts_filtered, all_posts_filtered)` at the end :crab: